### PR TITLE
[Merged by Bors] - feat(Analysis/NormedSpace/TrivSqZeroExt): The L1 norm 

### PR DESCRIPTION
**Pull request by eric-wieser**
The motivation here is to enable using `exp_add_of_commute` on dual numbers.

[This Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Exponentials.20in.20seminormed.20algebras/near/321981384) has some discussion about other possible norms.

---

- [x] depends on: #9491
- [x] depends on: #9796

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/NormedSpace/TrivSqZeroExt.lean [hunk 1]
```diff
@@ -188,6 +172,94 @@ theorem eq_smul_exp_of_ne_zero (x : tsze R M) (hx : x.fst ≠ 0) :
...
+    dsimp [fst, snd]
+    rw [add_assoc]
+    gcongr
+    · exact norm_mul_le _ _
+    refine (norm_add_le _ _).trans ?_
+    gcongr
+    · exact norm_smul_le _ _
+    refine (_root_.norm_smul_le _ _).trans ?_
+    rw [← add_zero (_ * _), mul_comm, MulOpposite.norm_op]
+    gcongr
```
Comment (eric-wieser) [inline Mathlib/Analysis/NormedSpace/TrivSqZeroExt.lean:223]: It would be very helpful if `gcongr` knew how to apply `norm_smul_le` and `norm_add_le` here.

**Comment by jcommelin** [comment]
> The motivation here is to enable using `exp_add_of_commute` on dual numbers.

Can you please remind me of the slightly bigger picture? Where is all this heading?

**Comment by eric-wieser** [comment]
There isn't really much bigger a picture here; I thought it would be nice to be able to use `NormedSpace.exp` on `DualNumber Real`, and the API for `NormedSpace.exp` is pretty unusable without a `NormedAlgebra` structure.

File: Mathlib/Analysis/NormedSpace/TrivSqZeroExt.lean [hunk 2]
```diff
@@ -176,4 +176,90 @@ end Field
 
 end Topology
 
+/-!
+### The $ℓ^1$ norm on the trivial square zero extension
+
+This is not particularly canonical, but is sufficient to obtain a normed algebra, and thus
+to use `exp_add_of_commute`.
```
Comment (j-loreaux) [inline Mathlib/Analysis/NormedSpace/TrivSqZeroExt.lean:183]: There may be some argument that it is canonical. This isn't it, but related: for regular normed algebras, the ℓ¹-norm on the unitization is maximal with respect to all norms (making it into a `NormedAlgebra` satisfying `NormOneClass`) for which the embedding is an isometry. It's possible there is something similar for `TrivSqZeroExt`, but it's also possible they are not that closely related.

File: Mathlib/Analysis/NormedSpace/TrivSqZeroExt.lean [hunk 3]
```diff
@@ -176,4 +176,90 @@ end Field
...
+
+section Ring
+
+variable [NormedField 𝕜] [NormedRing R] [NormedAddCommGroup M]
+variable [NormedAlgebra 𝕜 R] [NormedSpace 𝕜 M] [Module R M] [Module Rᵐᵒᵖ M]
+variable [BoundedSMul R M] [BoundedSMul Rᵐᵒᵖ M] [SMulCommClass R Rᵐᵒᵖ M]
+variable [IsScalarTower 𝕜 R M] [IsScalarTower 𝕜 Rᵐᵒᵖ M]
+
+instance instL1NormedAddCommGroup : NormedAddCommGroup (tsze R M) :=
+  inferInstanceAs <| NormedAddCommGroup (WithLp 1 <| R × M)
```
Comment (j-loreaux) [inline Mathlib/Analysis/NormedSpace/TrivSqZeroExt.lean:196]: Should we be more careful about this and avoid abusing defeq by pulling this back through a composition of equivs? For instance see [Unitization.instUniformSpace](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/NormedSpace/Unitization.html#Unitization.instUniformSpace) and [Unitization.instBornology](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/NormedSpace/Unitization.html#Unitization.instBornology).

Also, I know that `WithLp` on products is careful to replace the topology with the product topology, but can you add an `example` verifying this for `TrivSqZeroExt`?

File: Mathlib/Analysis/NormedSpace/TrivSqZeroExt.lean [hunk 4]
```diff
@@ -176,4 +176,90 @@ end Field
...
+    rw [add_assoc]
+    gcongr
+    · exact norm_mul_le _ _
+    refine (norm_add_le _ _).trans ?_
+    gcongr
+    · exact norm_smul_le _ _
+    refine (_root_.norm_smul_le _ _).trans ?_
+    rw [← add_zero (_ * _), mul_comm, MulOpposite.norm_op]
+    gcongr
+    positivity
```
Comment (j-loreaux) [inline Mathlib/Analysis/NormedSpace/TrivSqZeroExt.lean:224]:
```suggestion
    rw [mul_comm, MulOpposite.norm_op]
    exact le_add_of_nonneg_right <| by positivity
```

File: Mathlib/Analysis/NormedSpace/TrivSqZeroExt.lean [hunk 5]
```diff
@@ -176,4 +176,90 @@ end Field
...
+  norm_one := by rw [norm_def, fst_one, snd_one, norm_zero, norm_one, add_zero]
+
+end Ring
+
+section CommRing
+
+variable [NormedField 𝕜] [NormedCommRing R] [NormedAddCommGroup M]
+variable [NormedAlgebra 𝕜 R] [NormedSpace 𝕜 M] [Module R M] [Module Rᵐᵒᵖ M] [IsCentralScalar R M]
+variable [BoundedSMul R M]
+variable [IsScalarTower 𝕜 R M] [IsScalarTower 𝕜 Rᵐᵒᵖ M]
```
Comment (j-loreaux) [inline Mathlib/Analysis/NormedSpace/TrivSqZeroExt.lean:241]: You don't need the `𝕜` variables here
```suggestion
variable [NormedCommRing R] [NormedAddCommGroup M]
variable [Module R M] [Module Rᵐᵒᵖ M] [IsCentralScalar R M]
variable [BoundedSMul R M]
```

File: Mathlib/Topology/Instances/TrivSqZeroExt.lean [hunk 6]
```diff
@@ -34,14 +34,23 @@ local notation "tsze" => TrivSqZeroExt
...
+    TopologicalSpace (tsze R M) :=
   TopologicalSpace.induced fst ‹_› ⊓ TopologicalSpace.induced snd ‹_›
 
-instance [T2Space R] [T2Space M] : T2Space (tsze R M) :=
+instance [TopologicalSpace R] [TopologicalSpace M] [T2Space R] [T2Space M] : T2Space (tsze R M) :=
   Prod.t2Space
 
+instance [UniformSpace R] [UniformSpace M] : UniformSpace (tsze R M) where
+  toTopologicalSpace := instTopologicalSpace
+  __ := instUniformSpaceProd
```
Comment (j-loreaux) [inline Mathlib/Topology/Instances/TrivSqZeroExt.lean:46]: This is subject to the same comment above. Should we avoid abuse of defeq here and pull back the uniform structure explicitly?

**Review by j-loreaux — COMMENTED** [review]
Can you also update the module docstring with a brief explanation of the norm we're placing on this type and why you want it? This occurs lower in the file, but it would be nice to have it at the topic.

File: Mathlib/Analysis/NormedSpace/TrivSqZeroExt.lean (hunk 1 above)
Comment (j-loreaux) [reply Mathlib/Analysis/NormedSpace/TrivSqZeroExt.lean:223]: If I'm not mistaken, you're pining for `grw` for which there's an open PR: #8167

File: Mathlib/Analysis/NormedSpace/TrivSqZeroExt.lean (hunk 3 above)
Comment (eric-wieser) [reply Mathlib/Analysis/NormedSpace/TrivSqZeroExt.lean]: I think the defeq abuse is fine here. I would be more open to using the "right" way of doing things if it actually kept the nice defeqs, but without some of the constructions in #8961 I think that's currently not the case (which is why you have to prove a bunch of things to copy the structure, while I don't).

**Comment by j-loreaux** [comment]
@eric-wieser did you intend for this to still be `awaiting-author`?

**Comment by eric-wieser** [comment]
I did, but as the author I had forgotten that I did that. I had intended to come back and show `inr` and `inl` are isometries; now done with `simp` lemmas. I figure there's not much point bundling them.

**Comment by j-loreaux** [comment]
bors merge

**Pull request closed without merging**